File: Core/src/main/java/com/github/ezauton/core/actuators/implementations/BaseSimulatedMotor.java
package com.github.ezauton.core.actuators.implementations;

import com.github.ezauton.core.actuators.VelocityMotor;
import com.github.ezauton.core.localization.sensors.TranslationalDistanceSensor;
import com.github.ezauton.core.utils.Clock;
import com.github.ezauton.core.utils.Stopwatch;

import java.util.concurrent.TimeUnit;

/**
 * Describes a simulated motor with an encoder. The motor has infinite acceleration
 */
public class BaseSimulatedMotor implements VelocityMotor, TranslationalDistanceSensor {
    private final Stopwatch stopwatch;

    /**
     * Assumed to be in dist/second
     */
    protected double velocity = 0;
    private VelocityMotor subscribed = null;
    private double position = 0;

    /**
     * Create a basic simulated motor
     *
     * @param clock The clock to keep track of time with
     */
    public BaseSimulatedMotor(Clock clock) {
        this.stopwatch = new Stopwatch(clock);
    }

    /**
     * @return The motor to which the velocity is being applied
     */
    public VelocityMotor getSubscribed() {
        return subscribed;
    }

    /**
     * Change the motor to which the velocity will be applied
     *
     * @param subscribed The new motor instance
     */
    public void setSubscribed(VelocityMotor subscribed) {
        this.subscribed = subscribed;
    }

    /**
     * @param targetVelocity The target speed for the motor to be ran at
     */
    @Override
    public void runVelocity(double targetVelocity) {
        stopwatch.resetIfNotInit();
        if (subscribed != null) {
            subscribed.runVelocity(targetVelocity);
        }
        double popped = stopwatch.pop(TimeUnit.SECONDS);
        position += velocity * popped;
        this.velocity = targetVelocity;
    }

    @Override
    public double getPosition() {
        stopwatch.resetIfNotInit();
        position += velocity * stopwatch.pop(TimeUnit.SECONDS); // Convert millis to seconds
        return position;
    }

    @Override
    public double getVelocity() {
        return velocity;
    }
}

File: Core/src/main/kotlin/com/github/ezauton/core/actuators/implementations/BaseSimulatedMotor.kt
package com.github.ezauton.core.actuators.implementations

import com.github.ezauton.conversion.Distance
import com.github.ezauton.conversion.LinearVelocity
import com.github.ezauton.conversion.Units
import com.github.ezauton.conversion.meters
import com.github.ezauton.core.actuators.VelocityMotor
import com.github.ezauton.core.localization.sensors.TranslationalDistanceSensor
import com.github.ezauton.core.utils.Clock
import com.github.ezauton.core.utils.Stopwatch

/**
 * Describes a simulated motor with an encoder. The motor has infinite acceleration
 */
class BaseSimulatedMotor
/**
 * Create a basic simulated motor
 *
 * @param clock The clock to keep track of time with
 */
  (clock: Clock) : VelocityMotor, TranslationalDistanceSensor {
  private val stopwatch: Stopwatch = Stopwatch(clock)

  /**
   * Assumed to be in dist/second
   */
  override var velocity: LinearVelocity = Units.mps(0.0)
    private set
  /**
   * @return The motor to which the velocity is being applied
   */


  /**
   * The motor to which the velocity will be applied
   *
   */
  var subscribed: VelocityMotor? = null
  override var position = 0.0.meters
    get(): Distance {
      stopwatch.resetIfNotInit()
      field += velocity * stopwatch.pop()
      return field
    }
    private set

  /**
   * @param targetVelocity The target speed for the motor to be ran at
   */
  override fun runVelocity(targetVelocity: LinearVelocity) {
    stopwatch.resetIfNotInit()
    if (subscribed != null) {
      subscribed!!.runVelocity(targetVelocity)
    }
    val popped = stopwatch.pop()
    position += velocity * popped
    this.velocity = targetVelocity
  }
}
